Amine dev

amine-zzr opened the pull request:
Add `_getline` function re-implementation of C `getline` function with some changes to fit project requirements, the function has: 
- Stack `buffer` of size `1024` to use fewer calls to system function `read`.
- Use of static variable to save stat of read content.
- Include implementing other reusable functions like `strdup`, `realloc`, `strncat`...

Add also `strtok` function.

Co-authored-by: @rakati <ouhaddounouredine@gmail.com>

amine-zzr committed:
check this

amine-zzr committed:
mod

amine-zzr committed:
memory leak is checked

amine-zzr committed:
a simple implementation of the strtok C function

amine-zzr committed:
betty styled the _getline function

amine-zzr committed:
the realloc function Betty styled

amine-zzr committed:
memcpy function

rakati commented on _getline.c:
```diff
@@ -0,0 +1,52 @@
+ * Return: number of characters read, or -1 on failure.
+ */
+
+ssize_t _getline(char **lineptr, size_t *n, int fd)
+{
+	char *new_ptr;
+	static size_t count;
```
- Try to use this signature `int _getline(char **line, const int fd)`, 
- I can't see why you use `static count var`, the static variable will contain the remaining characters you read and you don't need them until the next calls for the `_getline` function.
- Avoid prefixing or suffixing pointers with the `ptr` word, use only the base name of the var.

amine-zzr replied:
i didn't understand you point about the `static var`

rakati replied:
I think you may understand from this pseudo code:
```c
int _getline(char *line, const int fd)
{
    int r, pos;
    static char *rd = NULL; // content read from fd
    char *tmp;
    char buff[BUFF_SIZE + 1];
  
    // Step 1 can be wrapped into a function
    pos = str_find(rd, '\n'); // if rd is NULL or character not found str_find return -1 otherwise position of char
    if (pos >= 0)
    {
         *line = str_copy(rd, pos ); // copy content until newline
         tmp = rd;
         rd = strdup(rd + pos + 1);
         free(tmp);
         return(pos);
    }
    while ((r = read(fd, buff ,BUFF_SIZE)) != 0)
    {
         if (r < 0) return (-1);
         buff[r] = '\0';
         strjoin(rd, buff);
         if (str_find('\n', buff) || r < BUFF_SIZE)
              break;
    }
     pos =  str_find(rd, '\n'); // or str_find(rd, EOF);
     *line = str_copy(rd, pos ); // copy content until newline
     tmp = rd;
     rd = strdup(rd + pos + 1);
     free(tmp);
     return(pos);
}
```

rakati commented on _getline.c:
```diff
@@ -0,0 +1,52 @@
+			return (-1); /* error indicating the failure of malloc */
+	}
+
+	while ((count += read(fd, *lineptr + count, INIT_SIZE)) != 0)
```
- It is better to use stack variables for the buffer, it will be faster than dynamic vars.
- define `BUFF_SIZE=1024`.

amine-zzr replied:
i just wanted to mimic the behavior of the standard getline(), i can't resize memory in stack on the runtime.
but if we define a buffer of size 1024, it will be enough to carry a line, is that what you meant ?

rakati replied:
no, we need to read the data fast and use fewer calls to read, so we will read 1024 chars, to be fast for read we will store data in the local string buffer, and the dynamic var will store only the remaining chars.
we don't need to resize stack var it is impossible, the stack var is used to read data using the read `function`.

rakati commented on _getline.c:
```diff
@@ -0,0 +1,52 @@
+			return (-1); /* error indicating the failure of malloc */
+	}
+
+	while ((count += read(fd, *lineptr + count, INIT_SIZE)) != 0)
```
here the count will not be 0, as you always add the reader values to `count` var that may already have a value.

amine-zzr replied:
okay

rakati commented on shell.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef SHELL_H
+#define SHELL_H
+
+#define GIGA (1 << 10)
```
what is the need for that?

amine-zzr replied:
forgot to remove it, but i will use it now for the buffer size

rakati commented:
> getline () and strtok () functions added

Add here a meaningful description for the commit, as all your changes will have this message as a one commit so make it descriptif.

amine-zzr committed:
new functions prototype added

amine-zzr committed:
the getline function reimplemented

amine-zzr committed:
some string manipulationg functions used by the getline function

amine-zzr committed:
not used functions deleted

amine-zzr committed:
process_line function prototype added

amine-zzr committed:
new function process_line added to rediuce _getline function number of lines
proccess_line function processes a line from the existing content, allocates memory to store the line, copies the content until the newline character

amine-zzr commented:
`getline` function utilizes a `buffer` of size 1024 to read data from the `stream`.
The `buffer` content is duplicated into dynamically allocated memory pointed to by a static pointer `rd` to preserve content for future calls from the `main` function.
The function dynamically allocates memory for a line pointer `line`.
It copies the content of `rd` until a newline character is encountered into the memory allocated for `line`.
The `rd` pointer is updated by removing the copied content, ensuring it points to the remaining content in the `buffer`.

rakati commented on _getline.c:
```diff
@@ -0,0 +1,97 @@
+
+int _getline(char **line, const int fd)
+{
+	static char *rd;
```
I think you have to initialize rd with `NULL`, otherwise it will contain an indeterminate value, and it will point to a random address that may invoke [Undefined behavior](https://en.wikipedia.org/wiki/Undefined_behavior).

amine-zzr replied:
static variables are initialized to NULL by default i think!!

rakati commented on _getline.c:
```diff
@@ -0,0 +1,97 @@
+	char buffer[SIZE];
+	int pos;
+
+	pos = (rd != NULL) ? (int)(_strchr(rd, '\n') - rd) : -1;
```
I think it will be good to do the handle NULL ptr in the `_strchr` function.

amine-zzr replied:
okey

rakati commented on _getline.c:
```diff
@@ -0,0 +1,97 @@
+
+	if (pos >= 0)
+	{
+		func(*line, &rd, pos);
```
is it a typo or did you define a function called `func`? if so try to use a meaningful name.

amine-zzr replied:
yes i forgot to change the name after writing the code

rakati commented on _getline.c:
```diff
@@ -0,0 +1,97 @@
+
+	if (pos >= 0)
+	{
+		func(*line, &rd, pos);
```
same here

rakati commented on _getline.c:
```diff
@@ -0,0 +1,97 @@
+		return (pos);
+	}
+
+	while ((pos = _read(fd, &rd, buffer)) >= 0)
+	{
+		if (pos > 0 || pos < SIZE)
+			break;
+	}
```
the conditions for breaking from the loop are:
- Newline found in the read content.
- An error occurred.
- reaching the end of the file.

amine-zzr replied:
check discord

amine-zzr committed:
handling the end of file or Ctrl + D

amine-zzr committed:
_strchr prototype modification

amine-zzr committed:
_strchr function returns (-1) if poinet passed is NULL, else, the position of the character passed as argument

rakati commented on _getline.c:
```diff
@@ -0,0 +1,106 @@
+{
+	char *tmp;
+
+	*line = malloc((pos + 1) * sizeof(char));
```
you forgot to protect malloc.

rakati commented on _getline.c:
```diff
@@ -0,0 +1,106 @@
+	char *tmp;
+
+	*line = malloc((pos + 1) * sizeof(char));
+	_strncpy(*line, *rd, pos);
```
I think if we have created functions similar to the C library at least we need to make sure they do similar things to use them in future projects, otherwise we give this function another name. 
please refer to the `man strncpy`.

rakati commented on _getline.c:
```diff
@@ -0,0 +1,106 @@
+
+	if (pos >= 0)
+	{
+		process_line(&*line, &rd, pos);
```
making the `process_line function` return an integer can reduce lines here, and make it easy to report errors from the `process_line` function.
- The `process_line` function should return a number of characters in a line (same as returning `pos`) or a negative number for an error.

rakati commented on _getline.c:
```diff
@@ -0,0 +1,106 @@
+		return (pos);
+	}
+
+	while ((pos = _read(fd, &rd, buffer)) >= 0)
```
can you use another variable than pos, each variable should have one function or job to do, this way everything will be easy to read and understand, try to declare an `int r` for the read return value.

rakati replied:
btw I got confused a bit using the same name read preceded with an underscore is tricky.
we proceed with a function name with an underscore for a re-implemented function by us but still do the same job as the original functions. please try to come up with a new name for the function `_read`.

rakati reviewed (changes requested):
Can you add tests for the `read` and  `strtok` functions? and please resolve conflicts.

amine-zzr committed:
test files

rakati committed:
change process_line to static function returns an int

amine-zzr reviewed (commented):
as we discussed in discord messages, you will take it from here.

amine-zzr committed:
simple strtok test

rakati committed:
rename func _strchr to _index

rakati committed:
rename func _strchr to _index in the header and improve func docstring

rakati committed:
change func call from _strchr to _index

rakati committed:
protect malloc, and improve getline doc

rakati committed:
adapt our _strncpy to the C strncpy

rakati committed:
Improve docs and fix betty style

rakati committed:
Use old condition in strncpy

rakati committed:
Handle erros in process_line func

rakati committed:
Solve conflict and re-order declarations in header

rakati committed:
rename _getline.c to getline.c

rakati committed:
mv string function to string_utility and rm old file

rakati committed:
mv _strtok and _strobrk funcs to parser file

rakati committed:
mv _index func to general_utility file

rakati committed:
adapt funcs declarations conflicts with defintions after merge

rakati committed:
Merge branch 'amine-dev' of github.com:rakati/simple_shell into amine-dev

rakati committed:
add const to protect variables and use BUFF_SIZE instead of SIZE

rakati committed:
implement realloc function

rakati committed:
adapt the getline test and example

rakati committed:
add const keyword to protect vars, and fix errata

rakati committed:
Fix several bugs
- implment realloc to realloc rd var befor concate buff with it.
- remove _read func, use less code and better handle conditions.
- fix bugs in process_line func, case of rd contains only one line
  avoid freeing or realloc-ing empty rd.

rakati committed:
Merge branch 'main' into amine-dev

Co-authored-by: rakati <ouhaddounouredine@gmail.com>

rakati reviewed (approved):
great job Amine, I fixed bugs present in the code, and now the test is working well with different `BUFF_SIZE`.

rakati committed:
make process_line func static to the file

Co-authored-by: rakati <ouhaddounouredine@gmail.com>

amine-zzr committed:
_strcmp prototype added

amine-zzr committed:
struct name changed to s_env and typedef to t_env as suggested

amine-zzr committed:
as a reasult we changed the prototypes of environ handeling functions

amine-zzr commented on getline.c:
```diff
@@ -0,0 +1,77 @@
+	*line = malloc(pos + 1);
+	if (*line == NULL)
+		return (-2);
+	*line = _strncpy(*line, *rd, pos);
+	if (*line == NULL)
+		return (-2);
```
what is the case for _strncpy to return NULL pointer?
